### DISCUSSION TITLE
[Feature] Can We support block method call when custom a plugin?

### DISCUSSION THREAD

**Discussion by kael-aiur** [apache/skywalking#11721] [Q&A] [answered]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

I'm trying to develop a java agent plugin which integrated sentinel to implement request degtade, when some request are trigger degrade logic,I need to prevent this call, but I found all of my exception are catch an just print a log in method interceptor, such as `org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstMethodsInter` or `org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstMethodsInterWithOverrideArgs`, Is it possible to interrupt the call by throwing a specific exception here?

### Are you willing to submit a pull request to implement this on your own?

- [X] Yes I am willing to submit a pull request on my own!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by wu-sheng** [comment, marked as answer]
If you throw an exception in your plugin codes, yes, it would be caught and printed out. There is no way to prevent that.
Because in Java, you never should use Exception to control the process flow, that costs high performance impact and doesn't follow Java best practice.

**Reply by kael-aiur** [reply]
I will throw an exception in my plugin, but I need the core code catch my exception an block this call,the exception will be defined in core code but never throw in core code, It just provides a way for the plug-in to interrupt the call

**Reply by wu-sheng** [reply]
If you want to block the call of the original method, use `MethodInterceptResult` in `#beforeMethod` method. Don't use an exception. Like I said, this should not be done, because if we provide that, several days later, you will come back and complain that your agent slows the service down.

**Reply by kael-aiur** [reply]
Okay, thank you for your guidance, but I still have a problem. I need to interrupt the entire call chain because I don't know what kind of object should be returned as the result of the downgrade. I want to throw this exception in the application to remind developers Is there an application degrade, or is there a better way to achieve this effect?

**Reply by wu-sheng** [reply]
No, there isn't. APM agent is monitoring oriented, I guess you are in another use case. We can't support that.

**Reply by kael-aiur** [reply]
Ok thanks, I will try other ways to implement my function

**Reply by kael-aiur** [reply]
you can close this feature, I have not other problem,thanks a lot.

**Comment by lujiajing1126** [comment]
For sure, you can do this. But you have to develop a customized `ClassEnhancePluginDefine`, let's say `ClassEnhancePluginDefineV3`. Then you need to develop `InstMethodsInterV3` to implement your own plugin workflow as a new instance/static method interceptor.

As you are talking about Sentinel, I would say it is complicated but possible. We've already integrated Sentinel with the SkyWalking Java Agent core but never share this to the community.

Besides, you may refer to AHAS agent provided by Alicloud as a commercial product. I think it could fit your need well.

**Reply by wu-sheng** [reply]
> For sure, you can do this. But you have to develop a customized ClassEnhancePluginDefine, let's say ClassEnhancePluginDefineV3. Then you need to develop InstMethodsInterV3 to implement your own plugin workflow in a new instance/static method interceptor.

That is possible. As it keeps on private rather than community responsibility, I think we are all good. :)

**Reply by kael-aiur** [reply]
Thank you for your suggestions. I will refer to your suggestions to re-evaluate my needs and technical solutions.

**Comment by kael-aiur** [comment]
> For sure, you can do this. But you have to develop a customized `ClassEnhancePluginDefine`, let's say `ClassEnhancePluginDefineV3`. Then you need to develop `InstMethodsInterV3` to implement your own plugin workflow as a new instance/static method interceptor.
> 
> As you are talking about Sentinel, I would say it is complicated but possible. We've already integrated Sentinel with the SkyWalking Java Agent core but never share this to the community.
> 
> Besides, you may refer to AHAS agent provided by Alicloud as a commercial product. I think it could fit your need well.

Thank you for your answer. You mentioned that you support Sentinel in the core code. Can I ask how you solve Sentinel's code dependencies? When executing the enhancement point of the method, due to class loading problems, I need to add Sentinel related classes to the application classpath. Is this how you solved it?

**Reply by chengasd** [reply]
We provide external with thin sdk like sentinel-core.jar, and introduce sentinel dependencies into Agent shade to other classpath. And adjust the intercept method in InstMethodsInterV2 by defining InstMethodsInterV3. When catch Sentinel BlockException，then handle circuit breaking.

**Reply by kael-aiur** [reply]
So the application does not need to modify the code, but needs to introduce the sdk and repackage it to enable the sentinel function based on skywalking?

**Reply by chengasd** [reply]
Yes, In our solution，it is like this. The application can implement customized degrade strategies through SPI. Of course, default strategies are also available.

**Reply by kael-aiur** [reply]
good idea! thanks for your answer.
